staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -388,20 +388,22 @@ func (w *watchCache) UpdateResourceVersion(resourceVersion string) {
 		return
 	}
 
-	w.Lock()
-	defer w.Unlock()
-	w.resourceVersion = rv
-
-	// Don't dispatch bookmarks coming from the storage layer.
-	// They can be very frequent (even to the level of subseconds)
-	// to allow efficient watch resumption on kube-apiserver restarts,
-	// and propagating them down may overload the whole system.
-	//
-	// TODO: If at some point we decide the performance and scalability
-	// footprint is acceptable, this is the place to hook them in.
-	// However, we then need to check if this was called as a result
-	// of a bookmark event or regular Add/Update/Delete operation by
-	// checking if resourceVersion here has changed.
+	go func() {
+		w.Lock()
+		defer w.Unlock()
+		w.resourceVersion = rv
```

Comment:
How does this prevent out of order execution? The prior code arguably might have been serialized but this code could end up with e.g. two copies of this goroutine both waiting on the lock, then the "wrong" one gets it first.

Reply:
This is relying on the reflector implementation that all Add/Update/Delete and UpdateResourceVersion are happening synchronously:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L493

Reply:
Updated the comment to reflect that.

Reply:
I can believe it for the below section, but not for this bit running in a goroutine.

1. Imagine 3 successive calls A, B, C resulting in a run of this goroutine.
2. Call A happens first, takes the lock, and then the CPU context-switches away.
3. Then call B starts and blocks waiting for the lock.
4. Then call C starts and blocks waiting for the lock.
5. Then call A gets another timeslice, completes, and releases the lock.
6. Then call **C** acquires the lock and sets RV
7. Then call **B** acquires the lock and sets RV, making it go back in time.

Reply:
OOps - thanks for catching this bug.
It wasn't supposed to be goroutine - it was supposed to be just func to lock only that instruction.

Fixed now - PTAL

Reply:
Ah that makes more sense :)